File: backend/app/schemas/topology.py
# PacketArch — OT Traffic Simulation Platform
# Copyright (c) 2026 Rocky Smith <rocky.d.smith@proton.me>
# Licensed under GPL-3.0. See LICENSE at the repo root.
"""Schemas for the multi-sensor topology ("Advanced Deployment") routes."""

from typing import Any

from pydantic import BaseModel, Field


class TopologyIssue(BaseModel):
    code: str
    message: str
    subject_id: str | None = None


class TopologySegment(BaseModel):
    span: str = Field(description="SPAN this frame appears on: 'zone:<zone_id>' or 'core'")
    src_mac: str
    dst_mac: str
    vlan: int | None = Field(default=None, description="802.1Q tag; None = untagged")
    ttl_delta: int = Field(description="IP TTL adjustment vs the generated packet (0 or -1)")


class TopologyFlowPlan(BaseModel):
    kind: str = Field(description="'intra' or 'cross'")
    source_zone: str
    target_zone: str
    segments_forward: list[TopologySegment]
    segments_reverse: list[TopologySegment]


class TopologyPreviewResponse(BaseModel):
    """Derived L1 topology + per-flow segment plans. Pure preview, no side effects."""

    valid: bool
    errors: list[TopologyIssue]
    warnings: list[TopologyIssue]
    switches: dict[str, dict[str, Any]] = Field(description="zone_id -> derived IE3500 switch")
    core: dict[str, Any] | None = Field(default=None, description="Derived IE9320 core with per-zone SVIs")
    links: list[dict[str, Any]] = Field(description="L1 links: device->switch (access), switch->core (trunk)")
    spans: list[dict[str, Any]] = Field(description="Capture SPANs: one per zone + core")
    flow_plans: dict[str, TopologyFlowPlan] = Field(description="flow_id -> segment plan")

File: frontend/src/api/topology.ts
/*
 * PacketArch — OT Traffic Simulation Platform
 * Copyright (c) 2026 Rocky Smith <rocky.d.smith@proton.me>
 * Licensed under GPL-3.0. See LICENSE at the repo root.
 */
/**
 * Multi-sensor topology ("Advanced Deployment") API — provisions one Local
 * Sensor Lab per zone + a core, with L1-aware per-segment traffic injection.
 */

import apiClient from './client';

export interface TopologyPreflight {
  scenario_id: string;
  sensor_count: number;
  ram_estimate_gb: number;
  spans: string[];
  switches: number;
  has_core: boolean;
  flow_plans: number;
}

export interface TopologyIssue {
  code: string;
  message: string;
  subject_id?: string | null;
}

export interface TopologyPreview {
  valid: boolean;
  errors: TopologyIssue[];
  warnings: TopologyIssue[];
  switches: Record<string, Record<string, unknown>>;
  core: Record<string, unknown> | null;
  links: Array<Record<string, unknown>>;
  spans: Array<Record<string, unknown>>;
  flow_plans: Record<string, unknown>;
}

export interface TopologyMember {
  span_id: string;
  role: string;
  lab_id: string;
  slug?: string | null;
  agent_id?: string | null;
  agent_token?: string | null;
  gen_if?: string | null;
  sensor_serial?: string | null;
}

export interface TopologyProvisionResult {
  scenario_id: string;
  sensor_count: number;
  ram_estimate_gb: number;
  members: TopologyMember[];
  span_interface_map: Record<string, string>;
}

export interface TopologyDeployment {
  scenario_id: string;
  sensor_count: number;
  members: Array<Record<string, unknown>>;
  torn_down?: Array<Record<string, unknown>> | null;
}

const base = (id: string) => `/api/v1/scenarios/${id}/topology`;

export const topologyApi = {
  preview: async (scenarioId: string): Promise<TopologyPreview> =>
    (await apiClient.post(`${base(scenarioId)}/preview`)).data,

  preflight: async (scenarioId: string): Promise<TopologyPreflight> =>
    (await apiClient.get(`${base(scenarioId)}/preflight`)).data,

  deploy: async (scenarioId: string): Promise<TopologyProvisionResult> =>
    (await apiClient.post(`${base(scenarioId)}/deploy`)).data,

  deployment: async (scenarioId: string): Promise<TopologyDeployment> =>
    (await apiClient.get(`${base(scenarioId)}/deployment`)).data,

  teardown: async (scenarioId: string): Promise<TopologyDeployment> =>
    (await apiClient.post(`${base(scenarioId)}/teardown`)).data,
};

export default topologyApi;
